### ISSUE TITLE
Services are not published into folders as expected

### ISSUE THREAD

**Issue by PleaseStopAsking** [Esri/arcgis-powershell-dsc#157]
When publishing services with the ArcGIS_Server_Service resource, it does not respect the folder key and instead publishes the services directly into the Site (root) folder. Along with not publishing the services into their respective folders, it throws an error.

```
ArcGIS_Server_Service bar {
  ServerHostName = $env:COMPUTERNAME
  PathToSourceFile = 'C:\Data\foo\bar.sd'
  ServiceName = 'bar'
  ServiceType = 'MapServer'
  Folder= 'bar_folder'
  State = 'STARTED'
  Ensure = 'Present'
  PublisherAccount = $arcgisAdminCredential
}
```

```
VERBOSE: [pkrvmcz6ng6k1lp]: LCM:  [ Start  Set      ]
VERBOSE: [pkrvmcz6ng6k1lp]: LCM:  [ Start  Resource ]  [[ArcGIS_Server_Service]bar]
VERBOSE: [pkrvmcz6ng6k1lp]: LCM:  [ Start  Test     ]  [[ArcGIS_Server_Service]bar]
WARNING: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] The names of some imported commands from the module 'ArcGISUtility' include unapproved verbs that might make them less discoverable. To find the commands with unapproved verbs, run the Import-Module command again with the Verbose parameter. For a list of approved verbs, type Get-Verb.
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] Check for existence of ServiceName:- bar ServiceType:- MapServer Folder:- bar_folder
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] Service with name 'bar' of type 'MapServer' not found in folder 'bar_folder'
VERBOSE: [pkrvmcz6ng6k1lp]: LCM:  [ End    Test     ]  [[ArcGIS_Server_Service]bar]  in 0.8040 seconds.
VERBOSE: [pkrvmcz6ng6k1lp]: LCM:  [ Start  Set      ]  [[ArcGIS_Server_Service]bar]
WARNING: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] The names of some imported comm
ands from the module 'ArcGISUtility' include unapproved verbs that might make them less discoverable. To find the commands
 with unapproved verbs, run the Import-Module command again with the Verbose parameter. For a list of approved verbs, type
 Get-Verb.
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] https://pkrvmcz6ng6k1lp:6443
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] Check for existence of ServiceName:- bar ServiceType:- MapServer Folder:- bar_folder
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] [DEBUG] Services:-
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] Publishing service ‘bar' of type 'MapServer' to 'bar_folder/bar/MapServer'
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] Item does not exist - uploading file C:\Data\foo\bar.sd
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] Uploaded item successfully i67b0e53d-c3a5-48bb-8d12-7a9a84dfa8fc
VERBOSE: [pkrvmcz6ng6k1lp]: [[ArcGIS_Server_Service]bar] [DEBUG] Job:- {"jobId":"jfb40454719aa4d59a6a5fb4808804e70","jobStatus":"esriJobSubmitted"}
VERBOSE: [pkrvmcz6ng6k1lp]: LCM:  [ End    Set      ]  [[ArcGIS_Server_Service]bar]  in 6.1480 seconds.

PowerShell DSC resource ArcGIS_Server_Service  failed to execute Set-TargetResource functionality with error message: 
Service failed to publish. Last Job Status Submitted. Executing... Executing (Publish Service Definition): 
PublishServiceDefinition i67b0e53d-c3a5-48bb-8d12-7a9a84dfa8fc # # Start Time: Wed May  1 20:38:23 2019 Getting server 
information (server folders, data stores, etc.). Validating service definition. WARNING 086222: Output directory in 
service definition is not set or is invalid. Using default output directory. ERROR 001369: Failed to create the service. 
Failed to execute (Publish Service Definition). Failed at Wed May  1 20:38:24 2019 (Elapsed Time: 0.18 seconds) Failed. 
    + CategoryInfo          : InvalidOperation: (:) [], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure
    + PSComputerName        : localhost
```

**Comment by PleaseStopAsking** [comment]
Further testing shows that if I run the configuration once more without any changes, it will successfully complete on the second run.

**Comment by PleaseStopAsking** [comment]
Reviewing calls from ArcGIS Server Manager when publishing an .sd file shows that the folder must be created as a separate task before publishing takes place within the Submit-ArcGISPublishJob function.

https://github.com/Esri/arcgis-powershell-dsc/blob/72c8a26007170cbdf231ab024501e99d2d82274b/Modules/ArcGIS/DSCResources/ArcGIS_Server_Service/ArcGIS_Server_Service.psm1#L621-L622

Also, the call to publish the .sd file is passing the `in_sdp_id` parameter only which is why the service is getting published to the ROOT directory and not into the specified folder. To publish to a given folder, I believe the parameter `in_config_overwrite` must be passed as well to indicate the folder such as the following (not tested).

```
{
    "folderName": "foo",
    "service": {
        ...
    }
}
```

**Comment by shailesh91** [comment]
@PleaseStopAsking were you able to get this to work?

**Comment by PleaseStopAsking** [comment]
@shailesh91 No, my team would not approve the time to fix this so I ended up building a python script (had an existing script to start from) to handle the publishing and am just calling it from a Script resource.

**Comment by cameronkroeker** [comment]
fixed in v3.1.0.
